html/modules/custom/reliefweb_subscriptions/src/ReliefwebSubscriptionsMailer.php

```diff
@@ -0,0 +1,2142 @@
+        'last' => $this->time->getRequestTime(),
+        'next' => $this->getNextSendingTime($subscription),
+      ];
+    }
+
+    $this->logger->info('Sent emails {sent} of {queued} queued notifications.', [
+      'sent' => $total_sent,
+      'queued' => count($notifications),
+    ]);
+
+    // Update the subscription logs.
+    $this->updateLogs($logs);
+  }
+
+  /**
+   * Queue notification.
+   *
+   * @param string $sid
+   *   Subscription id.
+   * @param array $options
+   *   Drush options.
+   *
+   * @command reliefweb_subscriptions:queue
+   * @usage reliefweb_subscriptions:queue
+   *   Queue emails.
+   * @validate-module-enabled reliefweb_subscriptions
+   * @option entity_type
+   *   Entity type.
+   * @option entity_id
+   *   Entity Id.
+   * @option last
+   *   Timestamp to use as the last time notifications were sent.
+   */
+  public function queue($sid, array $options = [
+    'entity_type' => '',
+    'entity_id' => 0,
+    'last' => 0,
+  ]) {
+    $this->logger->notice('Processing {sid}', [
+      'sid' => $sid,
+    ]);
+
+    $subscriptions = reliefweb_subscriptions_subscriptions();
+    if (!isset($subscriptions[$sid])) {
+      $this->logger->error('Invalid subscription id');
+      return;
+    }
+
+    $subscription = $subscriptions[$sid];
+
+    // Attempt to queue triggered notification.
+    if ($subscription['type'] === 'triggered') {
+      $entity_type = $options['entity_type'];
+      $entity_id = $options['entity_id'];
+
+      if (empty($entity_type) || empty($entity_id)) {
+        $this->logger->error('Invalid entity type or id');
+        return;
+      }
+
+      // Check if the entity exists.
+      $entity = $this->entityTypeManager
+        ->getStorage($entity_type)
+        ->load($entity_id);
+      if (empty($entity)) {
+        $this->logger->warning('Entity not found, skipping');
+        return;
+      }
+
+      $sids = [];
+      foreach ($subscriptions as $sid => $subscription) {
+        if ($subscription['type'] === 'triggered' && $subscription['bundle'] === $entity_type) {
+          // Check if the subscription trigger is met for the entity.
+          if (!isset($subscription['trigger']) || $subscription['trigger']($entity, $entity_type)) {
+            $sids[] = $sid;
+          }
+        }
+      }
+
+      // Skip if there are no subscriptions with subscribers.
+      $sids = $this->getSubscriptionsWithSubscribers($sids);
+      if (empty($sids)) {
+        return;
+      }
+
+      // Skip if there are no new notifications to queue.
+      $sids = $this->getSubscriptionsNotYetQueued($sids, $entity_type, $entity_id);
+      if (empty($sids)) {
+        return;
+      }
+
+      $notifications = [];
+      $notifications[] = [
+        'sid' => $sid,
+        'bundle' => $entity->bundle(),
+        'entity_id' => $entity_id,
+      ];
+
+      // Queue notifications for all the subscriptions.
+      $this->queueNotifications($notifications);
```

Comment:
Looks like a duplicate of what `queueTriggered` does. Probably `queueTriggered` should be called here instead.

Reply:
Moved